### PR TITLE
A few fixes in cpp parsing and formatting in some corner cases.

### PR DESCRIPTION
**Pull request by Dandielo**
I've started using Uncrustify and found a few issues in my codebase which I hopefully have properly fixed.
The original test are passing, and I've only extended the test cases adding new ones that fail without the fixes.

Here is the current list of the fixes:
* Fixes '&' reference seen as 'BYREF' in 'ARITH' context.
* Fixes lambda wrongly indented if called and not assigned to anything.
* Fixes ')' wrongly indented when 'noexcept' part of the signature.
* Fixes '&' reference seen as 'ARITH' in trailing return 'BYREF' context.

### CODE REVIEW

**Comment by Dandielo** [comment]
Looking through the issues this might close / fix the following: 
* re-opened #3116
* #2205, #3038
* #3284

File: src/combine.cpp [hunk 1]
```diff
@@ -1485,7 +1485,7 @@ void do_symbol_check(chunk_t *prev, chunk_t *pc, chunk_t *next)
          {
             set_chunk_type(pc, CT_ARITH);
 
-            if (chunk_is_token(prev, CT_WORD))
+            if (chunk_is_token(prev, CT_WORD) && !chunk_is_token(next, CT_NUMBER))
```
Comment (gmaurel) [inline src/combine.cpp:1488]: I preffer:
```.cpp
            if (  chunk_is_token(prev, CT_WORD)
               && !chunk_is_token(next, CT_NUMBER))
```

File: src/indent.cpp [hunk 2]
```diff
@@ -1585,7 +1585,8 @@ void indent_text(void)
 
             chunk_t *head     = chunk_get_prev_nc_nnl_np(frm.top().pc);
             chunk_t *tail     = nullptr;
-            bool    enclosure = frm.prev().pc != chunk_skip_to_match(frm.prev().pc);
+            chunk_t *frm_prev = frm.prev().pc;
+            bool    enclosure = frm_prev->parent_type != CT_FUNC_DEF && frm_prev != chunk_skip_to_match(frm_prev);
```
Comment (gmaurel) [inline src/indent.cpp:1589]: I preffer:
```.cpp
            bool    enclosure = (  frm_prev->parent_type != CT_FUNC_DEF
                                && frm_prev !=
chunk_skip_to_match(frm_prev));
```

**Review by gmaurel — CHANGES_REQUESTED** [review]
Also add a comment with the Issue #xxxx

**Comment by gmaurel** [comment]
Fine to see you working on uncrustify,
Thanks.

**Comment by Dandielo** [comment]
I've updated the code style, so it matches your preferences and added to each location the comment 'Issue #3407' representing this pull request.
If I misunderstood something, please don't hesitate to point it out.

Thanks!

**Review by gmaurel — APPROVED** [review]
Fine, thanks
